### ISSUE TITLE
don't unpack jars/zip files

### ISSUE THREAD

**Issue by fommil** [ensime/ensime-server#761]
currently we unpack files from archives and then tell clients about the unpacked version.

Two steps:
1. update emacs client to handle read-only loads of archived files (I'm sure this is supported)
2. update server to send URLs to the compressed files

**Comment by edani** [comment]
Third step: update emacs client (and protocol) to send URLs of compressed files when doing symbol lookups / sem highlight / etc.

**Comment by aemoncannon** [comment]
For reference, here's how cider-mode opens paths within archives 
https://github.com/clojure-emacs/cider/blob/e871b4c669bc887e2cdfec4445a0115c602f8d0d/cider-interaction.el#L536

**Comment by aemoncannon** [comment]
Interesting, looks like Emacs virtualizes file interactions with file-name-handler-alist: http://www.gnu.org/software/emacs/manual/html_node/elisp/Magic-File-Names.html

**Comment by aemoncannon** [comment]
We would need to agree on a file name format for files-within-jars. Looks like cider does:
^(jar|zip):file:(.+)!/(.+)
Where I guess '!/' is the separator between jar path and compressed file path.

**Comment by fommil** [comment]
if it agrees with the Apache VFS naming conventions, then we're sorted! https://github.com/ensime/ensime-server/blob/master/src/main/scala/org/ensime/indexer/package.scala#L15

**Comment by aemoncannon** [comment]
So... why do we want to do this? : )  It occurs to me that it's kinda nice to keep the client ignorant of archives. I mean, what's the down side?

**Comment by fommil** [comment]
we never clean up after ourselves when we unextract, plus we have to do the file handling on the server side which is work that we'd rather not have to do.

**Comment by aemoncannon** [comment]
> clean up
Agree it's annoying.

> file handling
Do you mean the unpacking itself? It looks like we do it with minimal
effort (loc-wise). Even the elisp to support files in jars will probably be
longer.

I'm wondering if this is a place where we can take a little hit in
complexity to make clients live's easier.

**Comment by fommil** [comment]
yeah, low priority

**Comment by fommil** [comment]
I got stung by this today. Was looking in Gen.scala of scalacheck, then upgraded scalacheck and did a jump to source and was really confused why I could see methods that the compiler refused to see.

As a workaround, we should perhaps delete the dep-src directory on startup.

**Comment by fommil** [comment]
dag namnit, this caught me today when I was swapping versions of a backend library and I couldn't understand why my use of the (changing) API wasn't working.

**Comment by aemoncannon** [comment]
This seems like two separate issues. The entitled,  about unzipping or not
not zipping, and then the one you mention about stale sources. For the
latter, maybe unpack the sources into jar-name-scoped directories?

**Comment by ancane** [comment]
:+1:

**Comment by pschorf** [comment]
My issue with supporting subfolders came up again as well. Is the consensus to keep doing extraction on the server? In that case would you be open to a PR which supports extracting a subdirectory of the jar?

**Comment by fommil** [comment]
We definitely want to move away from server side extraction. I don't understand your usecase.

**Comment by pschorf** [comment]
We have an internal build tool which publishes zip artifacts with sources, but instead of the normal path structure, they look like:

src/com/foocorp/A.java
src/com/foocorp/baz/B.java

With tools like intellij, we work around this by using a URI like jar:file:/tmp/foo.jar!/src

I filed #977 a while ago to add something similar to ensime, but it got rolled into the larger refactor. I'm happy to help push either approach forward (URIs in the .ensime file, server extracts subdir) or helping support reading archives in the emacs client.

**Comment by fommil** [comment]
right, so you've got a prefix at the start of the source file which probably messes up our source resolver. Well, it wouldn't be the first time we had a workaround for an internal build tool... but I think we should implement the fix in https://github.com/ensime/ensime-server/blob/master/core/src/main/scala/org/ensime/indexer/SourceResolver.scala

Start by writing a test in https://github.com/ensime/ensime-server/blob/master/core/src/it/scala/org/ensime/indexer/SourceResolverSpec.scala and the fix should be pretty easy.

But since this is a workaround for irritating, non-conforming, complexity elsewhere (and I'd really rather you fixed your build tool, but I appreciate that is not always possible due to politics) then add a descriptive comment. Basically if I come across batshit crazy code that doesn't have a comment, I assume the author didn't know what they were doing and delete it, but if there is a "I'm sorry I had to write this, but..." comment, I'll keep it if its tested.

**Comment by fommil** [comment]
this would allow us to remove the evil workaround in https://github.com/ensime/ensime-server/pull/1501

**Comment by fommil** [comment]
LineSourcePositionHelper

**Comment by fommil** [comment]
Oh, @hardmettle I just realised that Path is far better than URI here. I'll help by doing all the protocol stuff, let me know when you've got an the core code compiling

**Comment by hardmettle** [comment]
@fommil Do I have to revert it back to Path now? I am in the middle of fixing the compilation for URI change ?

**Comment by fommil** [comment]
Yeah sorry, but it should be more or less the same code. I just forgot we could use Java 7!

**Comment by hardmettle** [comment]
@fommil please take a look.

**Comment by fommil** [comment]
Cool! I will definitely take a good look tomorrow, but now is my bedtime.

**Comment by fommil** [comment]
I had a look through. This looks trickier than I thought it would be. Everywhere we assume the Path is a File needs to be treated with care

**Comment by hardmettle** [comment]
When we're using Path instead of File .. what else could it be if it's not an actual file ? Client would always want to refer/look for source "File" isn't it?

**Comment by fommil** [comment]
It'll be a source file inside a jar file.

**Comment by hardmettle** [comment]
In that case how do I get the Jar or antly archive file path .. If I can have that then this can be handled

**Comment by fommil** [comment]
Um sure there is but I'm afraid this API is all new to me so it's a journey of discovery :-)

There is a helper class in java called Paths, it might have some useful methods. Or we can parse the string.

**Comment by hardmettle** [comment]
Actually my confusion is path which is the parameter to SourceFileInfo (path:Path, ...) is always a java/scala file which may or may not be in a jar or it could be a .jar file as well? I hope I am not sounding stupid here.

**Comment by fommil** [comment]
it could be something that looks like `jar:file:/path/to/archive.jar!/package/name/MyThing.scala`

**Comment by aemoncannon** [comment]
Still concerned about the client impact of this change. All clients will need to be able to open archive URIs. Worse, any client extensions that operate at the file path level will need to understand these URIs. 

What if we lazily extract to paths that include a Sha of the jar?

**Comment by fommil** [comment]
@aemoncannon yeah that's why we're just targeting 2.0 to give them time to catch up. Atom are really keen for this and I know recent Emacs has support for browsing archives too.

**Comment by aemoncannon** [comment]
@fommil I'm curious what does atom stand to gain?
Our immediate clients might catch up, but the plugin ecosystem won't.

**Comment by fommil** [comment]
Oh this would be huge for performance, especially on Windows. We're extracting source files like crazy, and I've spoken to quite a few people who get really confused why they can't browse the third party source files (they expected everything to be there, not just the files we extracted).

Which plugins are you thinking of? We could have a special emacs mode like `ensime-readonly-mode` or something so that not all hooks are enabled by default.

**Comment by hardmettle** [comment]
@fommil So the next step is what needs to be done if it's a jar file reference ? Should the absolute path of the file be extracted from the given reference and then be passed on as the file path?

**Comment by fommil** [comment]
no you'll need to split it up and pass it to the presentation compiler as an archive file with a path component.

**Comment by fommil** [comment]
actually I'm not 100% sure how the presentation compiler would want to see that kind of file, so understanding what to do is probably the next move. Ask on gitter.im/scala/contributing or scala-ide/scala-ide for people who understand that API (but bare in mind that nobody really does)

**Comment by fommil** [comment]
removing "Low Hanging Fruit" to indicate that this one turned out to be trickier than expected.

**Comment by hardmettle** [comment]
It'll take a while for me to complete it because as you mentioned.. it's a bit tricky and I would need to dive into compiler internals to some extent to get this functioning .

**Comment by fommil** [comment]
yup, fun stuff :smile: I'll jump in and send you a PR at some point to help with the protocols. But that shouldn't stop you being able to get `core/it:test` going green

**Comment by hardmettle** [comment]
Also I have another question  .. as you mentioned that in case of jar reference it gets passed on to the presentation compiler as jar and its path .. what do we expect the compiler to give us in return ?
I mean what is it that we want the compiler to "ideally" do for us with those two piece of information.

**Comment by fommil** [comment]
I don't think you'd need to get involved in the code that sends messages to or receives messages from the presentation compiler. You'd just need to make sure that the `AbstractFile` that you construct is suitable for consumption by the presentation compiler. You might need to add some additional testing around this, and adapt existing test expectations. e.g. in the presentation compiler specs (or BasicWorkflow), write a test that involves loading a file from the scala and java standard libraries and ask for a type at point somewhere (you might need to use a regex to ensure that you go to the correct location instead of hardcoding it).

**Comment by fommil** [comment]
@hardmettle I think you forgot to push your branch

**Comment by fommil** [comment]
@hardmettle ! Ping, can you please push your branch? Ok (expected) that it doesn't compile.

**Comment by hardmettle** [comment]
I will push tonight for sure .. I continued working on it without realising that haven't pushed the branch .. Apologies for the delay

**Comment by fommil** [comment]
Cool thanks. Alas my only hack time this week has been the last three days so it'll have to be next week before I get to this.

**Comment by hardmettle** [comment]
@fommil Pushed the branch . I was in middle of lot changes and it has some errors to be fixed , but I still pushed it. Will continue work to make core compile meanwhile you can join in when you get free.

**Comment by fommil** [comment]
I've cherry picked your WIP into my branch, I'll try to get a little bit more done today and tomorrow. just `git reset --hard fommil/jar-urls` before you do any more to avoid conflicts.

**Comment by fommil** [comment]
@hardmettle I've implemented almost all the methods on `RichEnsimeFile` now and added tests.

I didn't think initially that the converter to `AbstractFile` should live in `utils`, that feels more like a `core` thing... but I'm very coming round to the concept because it would keep all the `EnsimeFile` logic in a single place.

**Comment by fommil** [comment]
`scala-compiler` added to `util` so we can add a `.asScalaCompilerFile` method or something similar at that level and implement it separately.

**Comment by fommil** [comment]
I've got `core` compiling now, but with holes in places. We need to be careful about reconstructing `SourceFileInfo`s / `EnsimeFile`s from the `AbstractFile`s that the scala compiler gives us.

**Comment by hardmettle** [comment]
@fommil I pulled your changes and pushed the merged changes into my branch . Since spent too much time sorting out what changes you've introduced and what changes I have been doing it is all mixed up now . Now that I have your changes can you tell me where to start up again ? What things needs to be completed which are left over?

**Comment by fommil** [comment]
there are some unimplemented methods, but I guess the best thing to do would be to get the core integration tests compiling and running, then implement missing methods until they pass :smile:

we should probably write a few extra tests for this to make sure we're covered. Once `core` is working, I'll fix up the protocol.

I might drop in and out of this, so please remember to push your changes or (better yet) send me a PR to my branch.

**Comment by hardmettle** [comment]
The core is compiling ... So I guess next step is to get the it test compiling and running . Just to confirm this should work `sbt it:test-only` isn't it?

**Comment by fommil** [comment]
`sbt core/it:test`

**Comment by fommil** [comment]
(don't forget the `core/` bit or it'll try to compile the protocols and get into an infinite loop)

**Comment by hardmettle** [comment]
@fommil Made some changes in `it` tests to get them compiling and running . But there are still some tests failing ? Do you want me to commit the changes before we can work on passing all the tests?

**Comment by fommil** [comment]
Yes please always always commit your work, no matter how small, or we'll end up duplicating each other's efforts. I'm dipping in and out of this branch when i find the time. I might get some time this weekend.

**Comment by hardmettle** [comment]
I have pushed in my changes , you can sync your copy with it. Once you get time to take a look into this branch please review and share your thoughts about the additional work/tests needs to be added.

**Comment by fommil** [comment]
Cool thanks

**Comment by fommil** [comment]
ok, cherry picked your stuff thanks. I'll chip away at this tonight. I have an evil idea for the protocol... a system property that handles converting outbound messages to the legacy cache directory. Could ease migration of the clients.

**Comment by hardmettle** [comment]
Can we save the complicated part for peer programming on hackday? I would want to be part of it to understand it better?

**Comment by fommil** [comment]
yeah, I'll be there, not sure if we'll finish it but we can make a good stab at it! :-)

**Comment by fommil** [comment]
not much to do...

```
[info] *** 55 TESTS FAILED ***
[error] Failed tests:
[error] 	org.ensime.core.RichPresentationCompilerThatNeedsJavaLibsSpec
[error] 	org.ensime.core.SemanticHighlightingSpec
[error] 	org.ensime.intg.BasicWorkflow
[error] 	org.ensime.core.ImplicitAnalyzerSpec
[error] 	org.ensime.core.JavaCompilerSpec
[error] 	org.ensime.core.StructureViewBuilderSpec
[error] 	org.ensime.model.SourcePositionSpec
[error] 	org.ensime.intg.ImplicitsWildcardImports
[error] 	org.ensime.core.RichPresentationCompilerSpec
[error] 	org.ensime.intg.DebugTest
```

err

**Comment by hardmettle** [comment]
:( I know. .But I am guilty too. .couldn't give much time to it.Focused only on getting through compilation

**Comment by fommil** [comment]
this branch needs a rebase

**Comment by hardmettle** [comment]
So are you rebasing your branch? And then I can pull from yours

**Comment by fommil** [comment]
yeah, I'll try to it tonight or tomorrow.

**Comment by fommil** [comment]
ok, I rebased. You can `git fetch -a ; get reset --hard fommil/jar-urls`

**Comment by fommil** [comment]
I didn't get to this, I was fixing other things that broke. And I need to rebase again as a result.

**Comment by hardmettle** [comment]
Should I rebase because I was looking into it though I didn't make any changes?

**Comment by hardmettle** [comment]
I have pushed some more changes and fixed one more failing test. Only one more is left which is `DebugTest.scala` . That is WIP .

**Comment by fommil** [comment]
Awesome! Was that the not-so-simple-java-file-object?

**Comment by hardmettle** [comment]
Sadly not. I have kept best for the last.I am just quickly clearing off the ones which can be done without any additional big changes. I will soon get to the not so simple one.

**Comment by fommil** [comment]
@hardmettle I rebased to `2.0` and cherry picked your commit into my branch. I'm going to look into the `NotSoSimpleJavaFileObject` today

**Comment by fommil** [comment]
status:

```
[info] *** 9 TESTS FAILED ***
[error] Failed tests:
[error] 	org.ensime.core.JavaCompilerSpec
[error] 	org.ensime.intg.DebugTest
```

**Comment by fommil** [comment]
@hardmettle ok, done! `JavaCompilerSpec` now passes. I pushed to my branch. I'll now look at doing an evil hack in the protocol layer (and applying `CanonFile`) to keep all the clients happy if you could look at the failing debug test?

**Comment by hardmettle** [comment]
@fommil Sure , I was away for a while. And umm .. I had just this status before I took a break :\ . So what went wrong here? Can you just fix the ```JavaCompilerSpec``` with my changes?
```
[info] *** 5 TESTS FAILED ***
[error] Failed tests:
[error] 	org.ensime.intg.JavaWorkflow
[error] 	org.ensime.core.JavaCompilerSpec
[error] 	org.ensime.intg.DebugTest
[error] (core/it:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 831 s, completed 5 Sep, 2016 11:38:02 PM 
```

**Comment by fommil** [comment]
I think you're using the wrong branch. Make sure to reset the mine, I cherry picked all your changes and rebased. Should only be debugtest to fix.

**Comment by hardmettle** [comment]
You mean ```git fetch -a``` and then ```git reset --hard fommil/jar-urls``` right?

**Comment by hardmettle** [comment]
@fommil fixed all DebugTest but one. Not able to find what is going wrong in test ```trigger/continue``` . Can you please take a quick look ?

**Comment by fommil** [comment]
I'm at Scala World this weekend but I'll try to take a look

**Comment by hardmettle** [comment]
@fommil I tried getting around fixing the DebugTest but cant figure out why is there an issue with file path being different in `DebugBreakEvent` . Can you please take a look?

**Comment by hardmettle** [comment]
@fommil did you manage to take  a look? I tried getting around the problem but couldnt find a solution to as to why the file paths are not matching and hence the failure of the test.

**Comment by fommil** [comment]
I'm on holiday, I'll take a look in 3 weeks :-)
